src/node/vscode.ts

```diff
@@ -52,33 +53,25 @@ export class VscodeProvider {
     const vscode = await this.fork()
 
     logger.debug("setting up vs code...")
-    return new Promise<ipc.WorkbenchOptions>((resolve, reject) => {
-      const onExit = (code: number | null) => reject(new Error(`VS Code exited unexpectedly with code ${code}`))
-
-      vscode.once("message", (message: ipc.VscodeMessage) => {
-        logger.debug("got message from vs code", field("message", message))
-        vscode.off("error", reject)
-        vscode.off("exit", onExit)
-        return message.type === "options" && message.id === id
-          ? resolve(message.options)
-          : reject(new Error("Unexpected response during initialization"))
-      })
 
-      vscode.once("error", reject)
-      vscode.once("exit", onExit)
-
-      this.send(
-        {
-          type: "init",
-          id,
-          options: {
-            ...options,
-            startPath,
-          },
+    this.send(
+      {
+        type: "init",
+        id,
+        options: {
+          ...options,
+          startPath,
         },
-        vscode,
-      )
+      },
+      vscode,
+    )
+
+    const message = await this.onMessage<ipc.OptionsMessage>(vscode, (message): message is ipc.OptionsMessage => {
+      // There can be parallel initializations so wait for the right ID.
```

Comment:
Loading the page in two separate tabs is the main way I think. If we had some sort of collaborative mode it would probably be a bigger problem then.

Reply:
I guess in theory it could also happen if you refresh a page before it finishes loading to send off two almost-parallel requests although I don't think I've hit it that way before.

Reply:
> Loading the page in two separate tabs is the main way I think. If we had some sort of collaborative mode it would probably be a bigger problem then.

Shouldn't we prevent sending parallel `init`'s at the same time?

Reply:
Or actually, doesn't each websocket get it's own vscode process? Or is that only its own extension host?

Reply:
There's a single VS Code process for all websockets and an extension host for each websocket. Each websocket also gets its own management connection where we initialize all the services VS Code uses (like the file service, environment service, language initialization, and so on).

Reply:
The services are shared too though, come to think. Language stuff isn't though, and there are some other differences in the initialization like which folder to open up (depending on the query parameter for example).